python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,722 @@
+            src_candidates = []
+            shift_rot_matrix = None
+            # If this pattern is one that we matched on the past but we
+            # now want to skip it, we can do so here.
+            if not (pattern_skip_array is None) and \
+               np.any(pattern_skip_array == pattern_idx):
+                print("Skipping previously matched bad pattern %i..." %
+                      pattern_idx)
+                continue
+            # Grab the sources to attempt to create this pattern.
+            pattern = sorted_catalog[pattern_idx: np.min((pattern_idx +
+                                                          n_check, n_source)),
+                                     :3]
+
+            # Construct a pattern given the number of points we are using to
+            # create it. We also convert our tolerances to radians at this
+            # time.
+            (ref_candidates, src_candidates, shift_rot_matrix, cos_shift,
+             sin_rot) = self._construct_pattern_and_shift_rot_matrix(
+                 pattern, n_match, max_cos_shift, max_cos_rot_sq,
+                 max_dist_rad)
+
+            # If we didn't match enough candidates we continue to the next
+            # pattern.
+            if len(ref_candidates) < n_match:
+                continue
+
+            # Now that we know our pattern and shift/rotation are valid we
+            # store the the rotated versions of our test points for later
+            # use.
+            tmp_rot_vect_list = []
+            for test_vect in test_vect_list:
+                tmp_rot_vect_list.append(np.dot(shift_rot_matrix, test_vect))
+            tmp_rot_vect_list.append(pattern_idx)
+            rot_vect_list.append(tmp_rot_vect_list)
+
+            # Test for if we have enough rotations, which agree, or if we
+            # are in optimistic mode.
+            if self._test_rotation_agreement(rot_vect_list, max_dist_rad) < \
+               n_agree - 1:
+                continue
+
+            # Perform final verify and check that we have enough matches.
+            matches, distances = self._match_sources(
+                    source_catalog[:, :3], shift_rot_matrix, max_dist_rad)
+            shift = np.degrees(np.arccos(cos_shift)) * 3600.
+            if len(matches) >= min_matches:
+                self.log.debug("Succeeded after %i patterns." % pattern_idx)
+                self.log.debug("\tShift %.4f arcsec" % shift)
+                self.log.debug("\tRotation: %.4f deg" %
+                               np.degrees(np.arcsin(sin_rot)))
+                return matches, distances, pattern_idx, shift
+
+        self.log.warn("Failed after %i patterns." % (pattern_idx + 1))
+        return [], [], None, None
+
+    def _compute_test_vectors(self, source_catalog):
+        """Function to create a list of vectors we can use to test the
+        different rotations against.
+        ------------------------------------------------------
```

Comment:
Here and for all methods:
- Please document the returned data (in full detail).
- Any method that returns more than one or two values should return a pipe_base Struct or an instance of a custom class. That makes it easier to document what's going on, makes it easier to use the returned data correctly and harder to mess up by getting the order wrong. it also adds some flexibility in that you can return additional fields without breaking anything, and remove fields in a way that reliably breaks code that relied on them (instead of that code mis-assign returned values to the wrong variables and then failing later in "interesting" ways).